### DISCUSSION TITLE
Loop over arbitrary buffer in a Func

### DISCUSSION THREAD

**Discussion by Doubl3MD** [halide/Halide#7191] [Q&A] [answered]
I'm as confused as the title of my question. So I try to explain what I want to do. :)

Say I have an IMAGE buffer grayscale X,Y dimensioned.

I also have a PATTERN buffer Z,3 dimensioned.

I want for per each pixel of the IMAGE do a loop over the PATTERN buffer. 
Do some expression with the 3 values at position Z in this PATTERN buffer.

Then carry on to the next pixel in the IMAGE do the loop with the 3 values from the PATTERN buffer Z times for the whole PATTERN buffer , etc...

I'm stuck in the 'feed forward' way of thinking and for now I don't see a way to achieve this.
I try to pseudo code for extra input:

```
for X in IMAGE  : 
  for Y  in IMAGE  :

    for Z in PATTERN : // this part/loop is not obvious to me on how to achieve this.
      //Do tests and stuff with every IMAGE(X,Y) and every PATTERN(Z, 1 2 and 3);
```
more concrete
```
ImageParam p_image(type_of<uint32_t>(), 2, "p_image");
ImageParam p_pattern(type_of<uint32_t>(), 2, "p_pattern");

Func toRealize;
Var X,Y,Z;

toRealize(X,Y) = ?how?  //Loop over p_pattern(Z times) and do something with : p_pattern(Z, 1 2 or 3) and p_image(X,Y)

toRealize.realize({X,Y},target,param);

```

Hopefully it make sense to somebody with a hint.

**Comment by abadams** [comment, marked as answer]
You want an RDom and an update definition for this, something like:

```
toRealize(x, y) = 0; // base case
RDom r(p_pattern);
toRealize(x, y) = some function of p_image(x, y) and p_pattern(r.x, r.y);

toRealize.in().realize(...);
```

The .in() call is optional, but it gets a slightly better schedule for this sort of code. It puts the base case inside the loop over x, y.

**Reply by Doubl3MD** [reply]
Thx! This got me looping!
